### PR TITLE
Workspace crud

### PR DESCRIPTION
**Pull request by miroslavrs81**

### CODE REVIEW

File: src/app/workspace/workspace.service.ts [hunk 1]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
     const user = await this.userRepository.findOneBy({ email });
     return { userExists: user ? true : false };
   }
+  
+  public async createWorkspace(createWorkspaceDto: CreateWorkspaceDto, user: User ):
+  Promise<{projectName: string, settings: string, ownerId: number}> {
+    const workspace = { ...createWorkspaceDto, ownerId: user.id};
+    // const workspace = this.workspaceRepository.create(createWorkspaceDto);
+    return this.workspaceRepository.save(workspace);
```
Comment (stefanj122) [inline src/app/workspace/workspace.service.ts:123]: Here you probably want to await this save method and why return type is't Promise<Workspace>?

File: src/app/workspace/workspace.service.ts [hunk 2]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
...
+  }
+
+  async findAllWorkspaces(): Promise<Workspace[]> {
+    return await this.workspaceRepository.find();
+  }
+
+  async findOneWorkspace(id: number): Promise<Workspace> {
+    const workspace = await this.workspaceRepository.findOneBy({ id });
+    if ( workspace.deletedAt === null ) {
```
Comment (stefanj122) [inline src/app/workspace/workspace.service.ts:132]: This condition will be true if workspace is't deleted, you probably want to check if deletedAt is not null.

File: src/app/workspace/workspace.service.ts [hunk 3]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
...
+      throw new Error('Workspace with id ' + { id } + ' was deleted');
+    } else if (!workspace) {
+      throw new Error('Workspace not found');
+    }
+    return workspace;
+  }
+
+  async updateWorkspace(id: number, updateWorkspaceDto: CreateWorkspaceDto) {
+    return await this.workspaceRepository.update(id, updateWorkspaceDto);
```
Comment (stefanj122) [inline src/app/workspace/workspace.service.ts:141]: Only owner of the workspace can update and delete workspace and return from update method won't give any useful information to FE.

File: src/app/workspace/workspace.controller.ts [hunk 4]
```diff
@@ -60,4 +66,51 @@ export class WorkspaceController {
...
+
+  @UseGuards(AuthGuard('jwt'))
+  @Post()
+  async createWorkspace(@Body() createWorkspaceDto: CreateWorkspaceDto, 
+  @GetUser() user: User):  Promise<{projectName: string, settings: string, ownerId: number}>
+    {
+    return this.workspaceService.createWorkspace(createWorkspaceDto,  user);
+  }
+
+  @Get()
```
Comment (Branimir) [inline src/app/workspace/workspace.controller.ts:78]: This should be filtered by the current user

File: src/app/workspace/workspace.controller.ts [hunk 5]
```diff
@@ -60,4 +66,51 @@ export class WorkspaceController {
...
+  async findAllWorkspaces(): Promise<Workspace[]> {
+    return await this.workspaceService.findAllWorkspaces();
+  }
+
+  @Get('/:id')
+  async findOneWorkspace(
+    @Param('id', ParseIntPipe) id: string,
+  ): Promise<Workspace> {
```
Comment (Branimir) [inline src/app/workspace/workspace.controller.ts:86]: This should be filtered by the current user as well

File: src/app/workspace/workspace.controller.ts [hunk 6]
```diff
@@ -60,4 +66,51 @@ export class WorkspaceController {
...
+    return await this.workspaceService.findOneWorkspace(+id);
+  }
+
+  @UseGuards(AuthGuard('jwt'))
+  @Put('/:id')
+   async updateWorkspace(
+    @Param('id', ParseIntPipe) id: number,
+    @Body() updateWorkspaceDto: CreateWorkspaceDto,
+    @GetUser() user: User ) {
+    return await this.workspaceService.updateWorkspace(+id, updateWorkspaceDto);
```
Comment (Branimir) [inline src/app/workspace/workspace.controller.ts:96]: This action should be possible only if the current user is owner

File: src/app/workspace/workspace.controller.ts [hunk 7]
```diff
@@ -60,4 +66,51 @@ export class WorkspaceController {
...
+  }
+
+  @UseGuards(AuthGuard('jwt'))
+  @HttpCode(204)
+  @Delete('/:id')
+  async removeWorkspace(
+    @Param('id', ParseIntPipe) id: number, 
+    @GetUser() user: User) {
```
Comment (Branimir) [inline src/app/workspace/workspace.controller.ts:104]: This action should be possible only if the current user is owner

File: src/app/workspace/workspace.controller.ts [hunk 8]
```diff
@@ -60,4 +66,51 @@ export class WorkspaceController {
...
+    return await this.workspaceService.removeWorkspace(+id);
+  }
+
+  @UseGuards(AuthGuard('jwt'))
+  @Patch('/:id/restore')     
+  async restoreWorkspace(
+    @Param('id', ParseIntPipe) id: number,
+    @GetUser() user: User ) {
```
Comment (Branimir) [inline src/app/workspace/workspace.controller.ts:112]: This action should be possible only if the current user is owner

File: src/app/workspace/workspace.service.ts [hunk 9]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
...
   }
+  
+  public async createWorkspace(createWorkspaceDto: CreateWorkspaceDto, user: User ):
+  Promise<{projectName: string, settings: string, ownerId: number}> {
+    const workspace = { ...createWorkspaceDto, ownerId: user.id};
+    // const workspace = this.workspaceRepository.create(createWorkspaceDto);
+    return this.workspaceRepository.save(workspace);
+  }
+
+  async findAllWorkspaces(): Promise<Workspace[]> {
```
Comment (Branimir) [inline src/app/workspace/workspace.service.ts:126]: Filter by current user

File: src/app/workspace/workspace.service.ts [hunk 10]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
...
+      throw new Error('Workspace not found');
+    }
+    return workspace;
+  }
+
+  async updateWorkspace(id: number, updateWorkspaceDto: CreateWorkspaceDto) {
+    return await this.workspaceRepository.update(id, updateWorkspaceDto);
+  }
+
+  async removeWorkspace(id: number) {
```
Comment (Branimir) [inline src/app/workspace/workspace.service.ts:144]: Check if owner

File: src/app/workspace/workspace.service.ts [hunk 11]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
...
+      return await this.workspaceRepository
+      .createQueryBuilder('ws')
+      .softDelete()
+      .where('id = :id', { id })
+      .execute();
+  }
+
+  async restoreWorkspace(id: number) {
```
Comment (Branimir) [inline src/app/workspace/workspace.service.ts:152]: Check if owner

**Review by Branimir — CHANGES_REQUESTED** [review]

File: src/app/workspace/workspace.service.ts [hunk 12]
```diff
@@ -114,4 +115,46 @@ export class WorkspaceService {
     const user = await this.userRepository.findOneBy({ email });
     return { userExists: user ? true : false };
   }
+  
+  public async createWorkspace(createWorkspaceDto: CreateWorkspaceDto, user: User ):
+  Promise<{projectName: string, settings: string, ownerId: number}> {
+    const workspace = { ...createWorkspaceDto, ownerId: user.id};
+    // const workspace = this.workspaceRepository.create(createWorkspaceDto);
```
Comment (miroslavrs81) [inline src/app/workspace/workspace.service.ts:122]: ownerId should be just owner

**Review by Branimir — APPROVED** [review]

**Pull request merged**